### PR TITLE
#115 - deploy workflows and commands

### LINKED ISSUE #115: Deploy on release
I would like to sync Github releases with Heroku deploy. I think we could use this package: https://github.com/marketplace/actions/deploy-to-heroku

A topic to discuss: are we going to deploy on releases only for live? Or maybe deploy for every main pull for test too?

### PR DESCRIPTION
**Pull request by mtracz**
This PR contains workflows for beta and production deploys. Also Makefile file with helpful commands.

### How it works?
**deploy to production**
Deploy will be triggered on tag creation (e.g. on release)
via SSH we will connect to the remote server, fetch, pull branch (`$GITHUB_REF_NAME`) - it will be `main`, and run deploy script via make (already installed on the server).

WIP. Will be done after beta tests

**deploy to beta**
Deploy will be triggered on comment `!deploy beta`  in PR, also can be triggered manually on specified branch.

First we will check if author of comment is one of `OWNER, MEMBER, COLLABORATOR` (github.event.comment.author_association), then we will check if comment is from PR `github.event.issue.pull_request`, at the end we will check if comment is equal to `!deploy beta`.

When these requirements will be met, we will use Github API to extract branch name (issue_comment event doesn't provide such data), from PR URL (`github.event.issue.pull_request.url`) we get PR number, and fetch details about this PR via Github API. If PR has not state `open` we will stop workflow/job.

After branch extracting we will use Github API to write comment under deploy trigger comment - `"Deploying to beta! :rocket:"`

The next step is `deploy` job, which will be run only if `pr_commented` job has result `success` or workflow has been run manually.
Job will use SSH action to connect to the remote server via SSH and use make scripts to deploy app. Action will pass `BRANCH_NAME` env variable to the script.

If deploy will pass, we will use Github API to write comment `Branch **${{ env.BRANCH_NAME }}** deployed to beta! :tada:`

---
### Concurrency and environment
**prod**
```yml
concurrency:
  group: deploy-prod
  cancel-in-progress: false
  ```
The `deploy-to-prod` workflow is using concurrency group with name `deploy-prod`.
Only one workflow in progress is allowed, the next one will be pending and will wait to finish the first one.

The environment for `deploy` job is `production`

**beta**
```yml
concurrency:
  group: deploy-beta
  cancel-in-progress: true
  ```
The `deploy-to-beta` workflow is using concurrency group with name `deploy-beta`.
Only one workflow in progress is allowed, the next one will cancel current running workflow.

The environment for `deploy` job is `beta`

---
### TODO

- switch from password to private key in SSH login
- limit manual deploy for author who is not `OWNER, MEMBER, COLLABORATOR`
- move manual deploy to another workflow?
- check risk of evaluating comment (`github.event.comment.body`) in if statement https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#understanding-the-risk-of-script-injections 

---
### Limitations
The limitations of `issue_comment` event is that, workflow will be run on every PR or ISSUE comment, but will be skipped if requirements wont be met.

---
#### Example of interaction
image

---
closes #115

### CODE REVIEW

**Comment by mtracz** [comment]
test

**Comment by mtracz** [comment]
test

File: environment/scripts/post-deploy-actions.sh [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+
+# -e is for "automatic error detection", tell shell to abort any time an error occurred
+set -e
+
+php artisan migrate --force && \
+  php artisan route:cache &&
+  bash reload-config.sh
```
Comment (jsawo) [inline environment/scripts/post-deploy-actions.sh:8]: missing end-of-line character

File: environment/scripts/post-deploy-actions.sh [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+
+# -e is for "automatic error detection", tell shell to abort any time an error occurred
+set -e
+
+php artisan migrate --force && \
+  php artisan route:cache &&
```
Comment (jsawo) [inline environment/scripts/post-deploy-actions.sh:7]: missing `\`?

**Review by jsawo — CHANGES_REQUESTED** [review]

File: environment/scripts/post-deploy-actions.sh (hunk 1 above)
Comment (mtracz) [reply environment/scripts/post-deploy-actions.sh:8]: Right, added.

File: environment/scripts/post-deploy-actions.sh (hunk 2 above)
Comment (mtracz) [reply environment/scripts/post-deploy-actions.sh]: Added.

**Review by kamilpiech97 — DISMISSED** [review]

**Review by jsawo — APPROVED** [review]

File: .github/workflows/deploy-to-beta.yml [hunk 3]
```diff
@@ -0,0 +1,103 @@
+name: Deploy to beta on trigger
```
Comment (krzysztofrewak) [inline .github/workflows/deploy-to-beta.yml:1]: Maybe `./github/workflows/deploy/beta.yml` would be better?

Comment (mtracz) [reply .github/workflows/deploy-to-beta.yml]: Moved under `deploy` dir. Renamed also.

**Review by krzysztofrewak — APPROVED** [review]

**Review by krzysztofrewak — APPROVED** [review]

**Pull request merged**